keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+```
+
+A few curated security profiles can be defined to cover most common use cases and shared in the community, and helps use of Kubernetes:
+
+- Simplicity bringing up a secured cluster -- when a cluster admin specifies a security profile at cluster creation time,
+they can be assured that the cluster will provide the security and isolation properties they need,
+without them having to understand the low-level configuration used to achieve it.
+- The profile is sharable and reusable as it’s defined in a portable way
+- Enforced automatically and continuously
+
+Additionally, a Security Profile
+
+- Is versioned, to allow seamless upgrade of the cluster and Security Profile
+- Is highly customizable for particular needs
+- Is highly extendable with extra features
+
+An example if using security profile is
+
+```
+kubeadm init --security-profile=default-1.0-0
+```
+
+The command above will configure a secure cluster according to security profile `default-1.0-0`.
+
+As security profiles cover common/generic use cases,
+they are portable and reusable in most Kubernetes deployments,
+They are **NOT** defining policies specific to a particular Kubernetes deployment, like:
+
+- ResourceQuota
+- RBAC rules
+- Rules in NetworkPolicy for particular workloads
+
+The proposal also works with Config-as-Code systems,
+see the [section](#work-with-config-as-code-systems) below for details.
+
+### Proposed Security Profiles
+
+#### default
+
+Full example: [default-1.0-0](#the-default-profile)
+
+Minimum security practice to be compatible with most workloads.
+Best fit for sharing the cluster with trusted workloads.
+
+Enforcement:
+
+- etcd
+  - TLS required for communication
+  - Explicitly specify `--wal-dir`, and `--max-wals=0`
+- kube-apiserver
+  - TLS required for communication
+  - Audit logging must be explicitly configured
+  - No anonymous auth, basic auth (unset `--anonymous-auth`, `--basic-auth-file`)
+  - Required authorization-mode: Node, RBAC
+  - Excluded authorization-mode: ABAC, AlwaysAllow
+  - Disable insecure port (`--insecure-port=0`, unset `--insecure-bind-address`)
+  - No auth token file (unset `--token-auth-file`)
+  - Disable profiling (unset `--profiling`)
+  - Request timeout is explicitly set
+  - Use service account credentials (`--use-service-account-credentials=true`)
+  - Required admission controls: AlwaysPullImages, DenyEscalatingExec, NamespaceLifecycle, ResourceQuota, PodSecurityPolicy, ServiceAccount, NodeRestriction, EventRateLimit
+  - Excluded admission controls: AlwaysAdmit
+  - Required feature gates to be ON: AdvancedAuditing, RotateKubeletServerCertificate, RotateKubeletClientCertificate
+- controller-manager
+  - Explicitly set `--terminated-pod-gc-threshold`
+- kubelet
+  - Disable hostname override (unset `--hostname-override`)
+  - No privileged
+- addons
+  - No dashboard
+- file-permissions
+  - ... # correct setup of cert/key/config file permissions
+- PodSecurityPolicy
+  - No priviledged
```

Comment:
can you specify? I'm thinking this is: privileged: false, allowPrivilegeEscalation: false, MustRunAsNonRoot

Or is this stronger than what you were thinking for 'default'?

Reply:
I think it's stronger than what the `default` is proposed. I'm even thinking of removing `No privileged` in this profile. The `default` profile is designed to give the maximum compatibility with existing workloads. Add restrictions to block `privileged` will definitely break some workloads. So I think I need rephrase the goal for each security profiles, and add one more between `default` and `saas-multitenancy`.